server/models/user.js

```diff
@@ -37,22 +37,37 @@ UserSchema.plugin(uniqueValidator, {
 UserSchema.pre('save', function(next) {
   const password = this.password;
   if (password) {
-    bcrypt.hash(password, 10)
-      .then((hash) => {
+    bcrypt
+      .hash(password, 10)
+      .then(hash => {
         this.password = hash;
         next();
       })
-      .catch((err) => next(err))
+      .catch(err => next(err));
   } else {
     next();
   }
 });
 
+UserSchema.statics.authenticate = function(email, password, callback) {
+  User.findOne({ email }, function(error, user) {
+    if (error) {
+      return callback(error);
+    }
+    if (user && bcrypt.compareSync(password, user.password)) {
+      return callback(null, user);
+    } else {
+      const err = Error('Incorrect username and password combination.');
```

Comment:
I'd love to easily import a custom (kind of like a clone) of the mongoose error structure. We may build out a utility to do that or maybe Mongoose comes with something that allows us to hook into how it generates its errors?

Either way, for now, let's at least provide the `name` and `message`, in a similar structure. Something like this:

```
{
  error {
    message: 'Incorrect username and password combination.',
    name: 'AuthenticationError'
  }
}
```

AuthenticationError is unique to our system. We'll eventually, probably, want to build out some reference that contains different types of errors we serve up. For now, I'm not worried about that, as long as we're consistent with naming these custom errors.

Reply:
I like this idea.  I will look into mongoose custom errors, but for now I have updated to throw an error with the message/name from above.